manager/controlapi/ca_rotation.go

```diff
@@ -0,0 +1,216 @@
+)
+
+var minRootExpiration = helpers.OneYear
+
+// Creates and updates the api.RootRotation object and the cross-signed intermediate.
+// This function assumes that a root rotation is not already in progress, and that the root
+// cert and key to be rotated in or the external CAs in the cluster have already been validated
+func updateRootRotationObject(ctx context.Context, securityConfig *ca.SecurityConfig, cluster *api.Cluster, newRootCA ca.RootCA, version uint64) error {
+	var (
+		rootCert, rootKey, crossSignedCert []byte
+		newRootHasSigner                   bool
+		err                                error
+	)
+
+	rootCert = newRootCA.Certs
+	if s, err := newRootCA.Signer(); err == nil {
+		rootCert, rootKey = s.Cert, s.Key
+		newRootHasSigner = true
+	}
+
+	// we have to sign with the original signer, not whatever is in the SecurityConfig's RootCA (which may have an intermediate signer, if
+	// a root rotation is already in progress)
+	switch {
+	case len(cluster.RootCA.CAKey) > 0:
+		var oldRootCA ca.RootCA
+		oldRootCA, err = ca.NewRootCA(cluster.RootCA.CACert, cluster.RootCA.CACert, cluster.RootCA.CAKey, ca.DefaultNodeCertExpiration, nil)
+		if err == nil {
+			crossSignedCert, err = oldRootCA.CrossSignCACertificate(rootCert)
+		}
+	case !newRootHasSigner: // the original CA doens't have a key, and the new CA doesn't have a key...
+		return grpc.Errorf(codes.InvalidArgument, "rotating from one external CA to a different external CA is not supported")
+	default:
+		// We need the same credentials but to connect to the original URLs (in case we are in the middle of a root rotation already)
+		externalCA := securityConfig.ExternalCA().Copy()
+		var urls []string
+		for _, c := range cluster.Spec.CAConfig.ExternalCAs {
+			if c.Protocol == api.ExternalCA_CAProtocolCFSSL && (len(c.CACert) == 0 || bytes.Equal(c.CACert, cluster.RootCA.CACert)) {
+				urls = append(urls, c.URL)
+			}
+		}
+		externalCA.UpdateURLs(urls...)
+		crossSignedCert, err = externalCA.CrossSignRootCA(ctx, newRootCA)
+	}
+
+	if err != nil {
+		log.G(ctx).WithError(err).Error("unable to generate a cross-signed certificate for root rotation")
+		return grpc.Errorf(codes.Internal, "unable to generate a cross-signed certificate for root rotation")
+	}
+
+	cluster.RootCA.RootRotation = &api.RootRotation{
+		CACert:            rootCert,
+		CAKey:             rootKey,
+		CrossSignedCACert: crossSignedCert,
+	}
+	cluster.RootCA.LastForcedRotation = version
+	return nil
+}
+
+// If there is no desired CA key, ensure that there is at least one external CA that uses the desired CA cert, if
+// a desired CA cert is specified.  If the desired CA cert is not specified, and there is no currently root CA key,
+// ensure that the config continues to provide an external CA corresponding to the cluster CA cert.
+func validateExternalCA(cluster *api.Cluster) error {
+	config := cluster.Spec.CAConfig
+
+	if len(config.SigningCACert) > 0 && len(config.SigningCAKey) == 0 {
```

Comment:
Maybe we can have a `cluster.RemoteSigner()`? or `cluster.LocalSigner()`? I just dislike seeing `Len(bla) > 0` everywhere, but I guess this is fine.